### PR TITLE
Functions and API are added as separate components.

### PR DESCRIPTION
**Pull request by Gayarohanangajanan**

### CODE REVIEW

File: components/TypingDNAAPI/src/main/webapp/META-INF/classloading.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+
+<Classloading xmlns="http://wso2.org/projects/as/classloading">
+
+    <!-- Parent-first or child-first. Default behaviour is child-first.-->
+    <ParentFirst>false</ParentFirst>
+
+    <Environments>CXF3,Carbon</Environments>
+</Classloading>
```
Comment (ashensw) [inline components/TypingDNAAPI/src/main/webapp/META-INF/classloading.xml:9]: Add a new line.

File: components/TypingDNAAPI/src/main/webapp/WEB-INF/beans.xml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            <bean class="com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider">
+                <constructor-arg>
+                    <bean class="com.fasterxml.jackson.databind.ObjectMapper">
+                        <property name="serializationInclusion" value="NON_NULL" />
+                    </bean>
+                </constructor-arg>
+            </bean>
+        </jaxrs:providers>
+    </jaxrs:server>
+</beans>
```
Comment (ashensw) [inline components/TypingDNAAPI/src/main/webapp/WEB-INF/beans.xml:28]: Add a new line.

Comment (ashensw) [reply components/TypingDNAAPI/src/main/webapp/WEB-INF/beans.xml:28]: Check other files as well.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/gen/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/factories/ServerApiServiceFactory.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.wso2.carbon.identity.conditional.auth.typingdna.api.factories;
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/gen/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/factories/ServerApiServiceFactory.java:1]: Add license header?

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
+package org.wso2.carbon.identity.conditional.auth.typingdna.api.core;
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java:1]: Add license header?

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/impl/MeApiServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package org.wso2.carbon.identity.conditional.auth.typingdna.api.impl;
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/impl/MeApiServiceImpl.java:1]: Add license header? Check all the other new classes that you have implemanted.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml [hunk 6]
```diff
@@ -0,0 +1,133 @@
+swagger: '2.0'
+info:
+  description: >
+    This is the RESTful API for a user to manage his/her TOTP QR URLs/secrets in
+    WSO2 Identity Server. Going forward this will be the API for user portal as
+    well as external parties.
+  version: v1
+  title: WSO2 Identity Server - TOTP API Definition
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml:8]: Change the title accordingly.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml [hunk 7]
```diff
@@ -0,0 +1,133 @@
+swagger: '2.0'
+info:
+  description: >
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml:3]: Change the description accordingly.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml (hunk 6 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml:8]: Check all the places in the swagger.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    This is the RESTful API for a user to manage his/her TOTP QR URLs/secrets in
+    WSO2 Identity Server. Going forward this will be the API for user portal as
+    well as external parties.
+  version: v1
+  title: WSO2 Identity Server - TOTP API Definition
+  contact:
+    name: WSO2
+    url: 'http://wso2.com/products/identity-server/'
+    email: architecture@wso2.com
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml:12]:
```suggestion
    email: iam-dev@wso2.org
```

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.functions/pom.xml [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        <carbon.kernel.package.import.version.range>[4.5.1, 5.0.0)</carbon.kernel.package.import.version.range>
+        <imp.pkg.version.javax.servlet>[2.5.0, 3.0.0)</imp.pkg.version.javax.servlet>
+        <carbon.identity.framework.imp.pkg.version.range>[5.14.97, 6.0.0)
+        </carbon.identity.framework.imp.pkg.version.range>
+        <osgi.framework.imp.pkg.version.range>[1.7.0, 2.0.0)</osgi.framework.imp.pkg.version.range>
+        <osgi.service.component.imp.pkg.version.range>[1.2.0, 2.0.0)</osgi.service.component.imp.pkg.version.range>
+        <carbon.identity.conditional.auth.functions.imp.pkg.version.range>[1.1.6, 2.0.0)
+        </carbon.identity.conditional.auth.functions.imp.pkg.version.range>
+        <carbon.user.api.imp.pkg.version.range>[1.0.1, 2.0.0)</carbon.user.api.imp.pkg.version.range>
+        <identity.governance.imp.pkg.version.range>[1.4.90, 2.0.0)</identity.governance.imp.pkg.version.range>
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.functions/pom.xml:227]: Shall we move these to the parent pom?

File: pom.xml [hunk 10]
```diff
@@ -7,9 +7,23 @@
...
+    <modules>
+        <module>components/org.wso2.carbon.identity.conditional.auth.typingdna.api</module>
+        <module>components/org.wso2.carbon.identity.conditional.auth.typingdna.functions</module>
+    </modules>
+    <dependencyManagement>
     <dependencies>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.conditional.auth.typingdna</groupId>
+            <artifactId>identity-conditional-auth-typingdna-functions</artifactId>
+            <version>1.0.0-SNAPSHOT</version>
```
Comment (ashensw) [inline pom.xml:20]:
```suggestion
            <version>${project.version}</version>
```

File: pom.xml [hunk 11]
```diff
@@ -7,9 +7,23 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.conditional.auth.typingdna</groupId>
+            <artifactId>identity-conditional-auth-typingdna-api</artifactId>
+            <version>1.0.0-SNAPSHOT</version>
```
Comment (ashensw) [inline pom.xml:25]:
```suggestion
            <version>${project.version}</version>
```

Comment (ashensw) [inline pom.xml:25]:
```suggestion
            <version>${project.version}</version>
```

**Review by ashensw — CHANGES_REQUESTED** [review]

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/pom.xml [hunk 12]
```diff
@@ -0,0 +1,159 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/pom.xml:1]: Add license.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/pom.xml [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+        </dependency>
+    </dependencies>
+
+
+
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/pom.xml:158]: Remove unnecessary new lines.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/pom.xml [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+</project>
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/pom.xml:159]: Add a newline

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+     * This function is used to delete users' typing patterns that are
+     * saved in typingDNA.
+     *
+     * @throws IOException When error/mistakes in TypingDNA configuration for tenant.
+     */
+    public void deleteTypingPatterns() throws IOException {
+
+        // Getting username & tenant domain & generating userID.
+        String username = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername();
+        String tenantDomain = "carbon.super";
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java:55]: Is the tenant domain hardcoded here? Is this working for tenant use cases?

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * Function that is used to get whether TypingDNA is Enabled or not for the tenant.
+     *
+     * @return Response that contains a boolean value Enabled-True , else - False.
+     */
+    public TypingdnaEnableResponseDTO getEnabled() {
+
+        String tenantDomain = "carbon.super";
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java:104]: Is the tenant domain hardcoded here? Is this working for tenant use cases?

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+            String line;
+            while ((line = rd.readLine()) != null) {
+                res.append(line);
+                res.append('\r');
+            }
+            rd.close();
+        } catch (MalformedInputException e) {
+            log.error(e);
+        } catch (IdentityEventException e) {
+            log.error(e);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java:92]: Need to throw a proper exception from here and handle it in the places that use this method.

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java (hunk 15 above)
Comment (Gayarohanangajanan) [reply components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java:55]: No tenant is not hardcoded here. but below method used to retrieve tenant will give a null value if the tenant is carbon.super. so defined the string as carbon.super at initial . below lines will change the name if tenant from context is null

Comment (Gayarohanangajanan) [reply components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/api/core/TypingDNAServiceImpl.java:55]: yes i will work for tenant use cases

**Comment by Gayarohanangajanan** [comment]
https://github.com/wso2/product-is/issues/10179

**Review by ashensw — APPROVED** [review]

File: components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+  BasicAuth:
+    type: basic
+  OAuth2:
+    type: oauth2
+    flow: accessCode
+    authorizationUrl: https://localhost:9443/oauth2/authorize
+    tokenUrl: https://localhost:9443/oauth2/token
+
+host: localhost:9443
+basePath: t/{tenant-domain}/api/typingdna/v1.0
```
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.conditional.auth.typingdna.api/src/main/resources/typingdna.yaml:132]: Why is this API deviated from the original format `t/{tenant-domain}/api/server/<version>/<context>` or `t/{tenant-domain}/api/user/...`

**Pull request merged**